modules/git/submodule.go

```diff
@@ -34,46 +43,73 @@ func getRefURL(refURL, urlPrefix, parentPath string) string {
 		return ""
 	}
 
-	url := strings.TrimSuffix(refURL, ".git")
-
-	// git://xxx/user/repo
-	if strings.HasPrefix(url, "git://") {
-		return "http://" + strings.TrimPrefix(url, "git://")
-	}
+	refURI := strings.TrimSuffix(refURL, ".git")
 
-	// http[s]://xxx/user/repo
-	if strings.HasPrefix(url, "http://") || strings.HasPrefix(url, "https://") {
-		return url
+	prefixURL, _ := url.Parse(urlPrefix)
+	urlPrefixHostname, _, err := net.SplitHostPort(prefixURL.Host)
+	if err != nil {
+		urlPrefixHostname = prefixURL.Host
 	}
 
 	// Relative url prefix check (according to git submodule documentation)
-	if strings.HasPrefix(url, "./") || strings.HasPrefix(url, "../") {
+	if strings.HasPrefix(refURI, "./") || strings.HasPrefix(refURI, "../") {
 		// ...construct and return correct submodule url here...
 		idx := strings.Index(parentPath, "/src/")
 		if idx == -1 {
-			return url
+			return refURI
 		}
-		return strings.TrimSuffix(urlPrefix, "/") + parentPath[:idx] + "/" + url
+		return strings.TrimSuffix(urlPrefix, "/") + parentPath[:idx] + "/" + refURI
 	}
 
-	// sysuser@xxx:user/repo
-	i := strings.Index(url, "@")
-	j := strings.LastIndex(url, ":")
+	if !strings.Contains(refURI, "://") {
+		// scp style syntax which contains *no* port number after the : (and is not parsed by net/url)
+		// ex: git@try.gitea.io:go-gitea/gitea
+		match := scpSyntax.FindAllStringSubmatch(refURI, -1)
+		if len(match) > 0 {
 
-	// Only process when i < j because git+ssh://git@git.forwardbias.in/npploader.git
-	if i > -1 && j > -1 && i < j {
-		// fix problem with reverse proxy works only with local server
-		if strings.Contains(urlPrefix, url[i+1:j]) {
-			return urlPrefix + url[j+1:]
+			m := match[0]
+			refHostname := m[2]
+			path := m[3]
+
+			if !strings.HasPrefix(path, "/") {
+				path = "/" + path
+			}
+
+			if urlPrefixHostname == refHostname {
+				return prefixURL.Scheme + "://" + urlPrefixHostname + path
+			}
+			return "http://" + refHostname + path
```

Comment:
Shouldn't it default to https better?

Reply:
It wouldn't work if the server didn't support https (we don't know anything about it in this case). most servers should issue a redirect from http to https if they do support it.

Reply:
@lafriks what do you think? I agree with @mrsdizzie here - most https sites will simply redirect http to https.

Reply:
true but from security point of view it would be better to prefer https and as most of users should be running https this would result in unneeded http request to server.. So it is trade-off for security vs usability :)

Reply:
I think we might need to add some repository configuration to allow for people to set their own conversions for submodules, but it's not particularly high priority.